### PR TITLE
Deathmatch modifiers

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
This PR adds the base to simple modifiers that the host can select to make the minigame ~~worse~~ more entertaining for everyone.

Here's the screenshot of the UI (without a few modifiers I added later):
immagine

## Why It's Good For The Game
I've seen this minigame is frankly popular lately, so I thought I could contribute to it.

## Changelog

:cl:
add: Added 20+ modifiers to the deathmatch minigame, which can be enabled by the host.
/:cl:

### CODE REVIEW

**Comment by Admiral-Gigaton** [comment]
you should add a no-gravity mode

**Comment by Ghommie** [comment]
Good Idea, done.

**Comment by EricZilla** [comment]
Would it be possible to do wings or perhaps everyone is snails

File: code/modules/deathmatch/deathmatch_modifier.dm [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	SIGNAL_HANDLER
+	bonus_spread_values[MIN_BONUS_SPREAD_INDEX] += 10
+	bonus_spread_values[MAX_BONUS_SPREAD_INDEX] += 35
+
+/datum/deathmatch_modifier/four_hands
+	name = "Four Hands"
+	description = "When one pair isn't enough..."
+
+/datum/deathmatch_modifier/four_hands/apply(mob/living/carbon/player, datum/deathmatch_lobby/lobby)
+	player.change_number_of_hands(4)
```
Comment (Iamgoofball) [inline code/modules/deathmatch/deathmatch_modifier.dm:151]: does this even work still

**Comment by Admiral-Gigaton** [comment]
set species modifier? might be harder to do, but besides abductors communicating i dont see anything wrong here

oh wait nightmare jaunt too

**Comment by SomeoneYouProbablyKnow** [comment]
Holy moly, these kick a ridiculous amount of ass.

File: code/modules/deathmatch/deathmatch_modifier.dm (hunk 1 above)
Comment (Ghommie) [reply code/modules/deathmatch/deathmatch_modifier.dm:151]: Yeah, just tested it at the thunderdome.

File: code/__DEFINES/traits/declarations.dm [hunk 2]
```diff
@@ -555,6 +555,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 
 /// This mob is phased out of reality from magic, either a jaunt or rod form
 #define TRAIT_MAGICALLY_PHASED "magically_phased"
+/// flips the direction of client movement
+#define TRAIT_INVERTED_MOVEMENT "inverted_movement"
```
Comment (MrMelbert) [inline code/__DEFINES/traits/declarations.dm:559]: Funnily enough you can do this without a trait (IE, by using an element). 

You can hook `COMSIG_MOB_CLIENT_PRE_MOVE` and do `move_args[MOVE_ARG_DIRECTION] = REVERSE_DIR(move_args[MOVE_ARG_DIRECTION])`. We use this for the confusion status effect. 

You don't have to do that in this pr, but just a neat idea to reduce the surface area of this.

**Comment by MrMelbert** [comment]
Fun fact: You can transform a mob's game plane to flawlessly flip someone's entire view around. Just an idea.

File: code/__DEFINES/traits/declarations.dm (hunk 2 above)
Comment (Ghommie) [reply code/__DEFINES/traits/declarations.dm]: Done.

**Comment by Admiral-Gigaton** [comment]
X ray/thermal eyes modifier?

random augs modifier?

**Comment by Ghommie** [comment]
> X ray/thermal eyes modifier?
> 
> random augs modifier?

X-ray vision is already in. I could do thermals too, I guess; random augs perhaps.

**Comment by Admiral-Gigaton** [comment]
> > X ray/thermal eyes modifier?
> > random augs modifier?
> 
> X-ray vision is already in, but I could do thermals too, I guess; random augs, perhaps.

played a recent game of DM on Terry, an admin spawned a miner demon, maybe a random "mega" fauna or something could be spawned? might be OP, or round ruining, but you dont HAVE to enable it.

though that just sounds like bitrunning for ghosts. doesnt matter, just was a fun little thing fighting the demon and the other people, dodging ice attacks and getting shot in the mouth.

**Comment by Ghommie** [comment]
> played a recent game of DM on Terry, an admin spawned a miner demon, maybe a random "mega" fauna or something could be spawned? might be OP, or round ruining, but you dont HAVE to enable it.

lol, that was me fucking around a bit yesterday. The idea is nice, but I think that should belongs to a more complex system for stuff like 'gamemodes' rather than modifiers.

**Comment by Admiral-Gigaton** [comment]
what about a slow regen, or slow degeneration of health modifier?

i had an idea of random tiles getting temporarily set on fire or damage you, but that wouldnt be easy to implement, so what about a player slowdown/speed up for their running speed?

**Comment by jughu** [comment]
splitted skull modifier so anyone who slips loses their brain

File: code/modules/deathmatch/deathmatch_modifier.dm [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+///Apply the modifier to the newly spawned player as the game is about to start
+/datum/deathmatch_modifier/proc/apply(mob/living/carbon/player, datum/deathmatch_lobby/lobby)
+	return
+
+/datum/deathmatch_modifier/health
+	name = "Double-Health"
+	description = "Doubles your starting health"
+	var/multiplier = 2
+	blacklisted_modifiers = list(/datum/deathmatch_modifier/health/triple)
```
Comment (Jacquerel) [inline code/modules/deathmatch/deathmatch_modifier.dm:54]:
```suggestion
	blacklisted_modifiers = list(/datum/deathmatch_modifier/health/triple)
	var/multiplier = 2
```
New vars below old

File: code/modules/deathmatch/deathmatch_modifier.dm [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+	var/obj/item/implant/explosive/deathmatch/implant = new()
+	implant.implant(player, silent = TRUE, force = TRUE)
+
+/datum/deathmatch_modifier/helgrasp
+	name = "Helgrasped"
+	description = "Cursed hands are being thrown at you!"
+
+/datum/deathmatch_modifier/helgrasp/apply(mob/living/carbon/player, datum/deathmatch_lobby/lobby)
+	var/metabolism_rate = /datum/reagent/inverse/helgrasp/heretic::metabolization_rate
+	player.reagents.add_reagent(/datum/reagent/inverse/helgrasp/heretic, initial(lobby.map.automatic_gameend_time) / metabolism_rate)
```
Comment (Jacquerel) [inline code/modules/deathmatch/deathmatch_modifier.dm:348]: We really need to make this a status effect which is _also_ applied by the reagent istg
Not in this PR though

File: code/modules/deathmatch/deathmatch_modifier.dm [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+			continue
+		///don't spawn mine next to player spawns.
+		if(locate(/obj/effect/landmark/deathmatch_player_spawn) in range(1, target_turf))
+			continue
+		var/mine_path = pick(mines)
+		new mine_path (target_turf)
+
+/datum/deathmatch_modifier/flipping
+	name = "Perma-Flipping"
+	description = "You're constantly flipping - the mob that's - however it's purely cosmetic"
```
Comment (Jacquerel) [inline code/modules/deathmatch/deathmatch_modifier.dm:400]:
```suggestion
	description = "You're constantly flipping - the mob that's - however it's purely cosmetic"
```
I don't think I understand what the hypenated part of this sentence is supposed to say

File: code/modules/mob/living/basic/farm_animals/pony.dm [hunk 6]
```diff
@@ -43,7 +43,7 @@
 	AddElementTrait(TRAIT_WADDLING, INNATE_TRAIT, /datum/element/waddling)
 	AddComponent(/datum/component/tameable, food_types = list(/obj/item/food/grown/apple), tame_chance = 25, bonus_tame_chance = 15, after_tame = CALLBACK(src, PROC_REF(tamed)), unique = unique_tamer)
 
-/mob/living/basic/pony/proc/tamed(mob/living/tamer)
+/mob/living/basic/pony/tamed(mob/living/tamer, atom/food)
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/farm_animals/pony.dm:46]: Understand the desire but a little bit spooked by this standardisation.
I think if you are creating a contract that this _must_ be used then perhaps you should just remove the ability to pass a callback to the component and always call this proc specifically.

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/modules/deathmatch/deathmatch_modifier.dm (hunk 5 above)
Comment (Ghommie) [reply code/modules/deathmatch/deathmatch_modifier.dm:400]: Yeah, me neither. I must have been pretty distracted to write rubbish.

**Comment by Jacquerel** [comment]
You got conflicts

**Pull request merged**